common/minicurl.cc

```diff
@@ -170,3 +202,156 @@ bool MiniCurl::postURL(const std::string& url,
   }
   return retval;
 }
+
+// MiniCurlMulti
+
+MiniCurlMulti::MiniCurlMulti() : d_ccs(numMultiCurlConnections)
+{
+  d_mcurl = curl_multi_init();
+  initMCurl();
+}
+
+MiniCurlMulti::MiniCurlMulti(size_t num_connections) : d_ccs(num_connections)
+{
+  d_mcurl = curl_multi_init();
+  initMCurl();
+}
+
+MiniCurlMulti::~MiniCurlMulti()
+{
+  if (d_mcurl != nullptr) {
+    curl_multi_cleanup(d_mcurl);
+  }
+}
+
+void MiniCurlMulti::initMCurl()
+{
+#ifdef CURLPIPE_HTTP1
+  curl_multi_setopt(d_mcurl, CURLMOPT_PIPELINING,
+                    CURLPIPE_HTTP1 | CURLPIPE_MULTIPLEX);
+#else
+  curl_multi_setopt(d_mcurl, CURLMOPT_PIPELINING, 1L);  
+#endif
+  curl_multi_setopt(d_mcurl, CURLMOPT_MAX_HOST_CONNECTIONS, d_ccs.size());
+  d_current = d_ccs.begin();
+}
+
+bool MiniCurlMulti::addPost(unsigned int id, const std::string& url,
+                            const std::string& post_body,
+                            const MiniCurlHeaders& headers)
+{
+  if (d_mcurl) {
+    // return false if we don't have any spare connections
+    if (d_current == d_ccs.end())
+      return false;
+    d_current->setPostData(url, post_body, headers);
+    d_current->setID(id);
+    curl_multi_add_handle(d_mcurl, d_current->getCurlHandle());
+    ++d_current;
+    return true;
+  }
+  return false;
+}
+
+const std::vector<mcmPostReturn> MiniCurlMulti::runPost()
+{
+  std::vector<mcmPostReturn> post_ret;
+  if (d_mcurl) {
+    int still_running;
+    int without_fds = 0;
+    do {
+      CURLMcode mc;
+      int numfds;
+      mc = curl_multi_perform(d_mcurl, &still_running);
+      if (mc == CURLM_OK) {
+        auto start_time = std::chrono::steady_clock::now();
+        
+        mc = curl_multi_wait(d_mcurl, NULL, 0, 1000, &numfds);
+
+        auto wait_time_ms = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start_time);
+        
+        if (mc == CURLM_OK) {
+          struct CURLMsg *m;
+          do {
```

Comment:
this could be easier to read with `while((m = curl_multi_info_read(d_mcurl, &msgq)) != NULL) {`